src/Tools/BuildValidator/Program.cs

```diff
@@ -306,30 +321,33 @@ private static bool ValidateFiles(IEnumerable<FileInfo> originalBinaries, Option
                 if (!referenceResolver.TryResolveReferences(metadataReferenceInfos, out var metadataReferences))
                 {
                     logger.LogError($"Failed to rebuild {originalBinary.Name} due to missing metadata references");
-                    return CompilationDiff.CreatePlaceholder(originalBinary, isError: true);
+                    return CompilationDiff.CreateMissingReferences(assemblyInfo, referenceResolver, metadataReferenceInfos);
                 }
                 logResolvedMetadataReferences();
 
                 var sourceLinks = ResolveSourceLinks(optionsReader, logger);
                 if (sourceResolver.ResolveSources(sourceFileInfos, sourceLinks, encoding) is not { } sources)
                 {
                     logger.LogError($"Failed to resolve sources");
-                    return CompilationDiff.CreatePlaceholder(originalBinary, isError: true);
+                    return CompilationDiff.CreateMiscError(assemblyInfo, "Failed to resolve sources");
                 }
                 logResolvedSources();
 
-                var (compilation, isError) = buildConstructor.CreateCompilation(
-                    optionsReader,
-                    originalBinary.Name,
-                    sources,
-                    metadataReferences);
-                if (compilation is null)
+                Compilation? compilation = null;
```

Comment:
>Compilation? compilation = null [](start = 16, length = 31)

Minor: Looks like this can be declared non-nullable with no default value: `Compilation compilation;`